### ISSUE TITLE
XMPP client using OAUTH2 mechanism crashes ejabberd 17.01

### ISSUE THREAD

**Issue by whtan98** [processone/ejabberd#1581]
> What version of ejabberd are you using?

17.01

> What operating system (version) are you using?

 Ubuntu 16.04.1 LTS

> How did you install ejabberd (source, package, distribution)?

Download from ProcessOne website

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

Send the following stanza cause ejabberd to crash:

INPUT
```
# XML Stanza
<auth xmlns="urn:ietf:params:xml:ns:xmpp-sasl" mechanism="X-OAUTH2">AGFkbWluQHhtcHAubWF0aW9uYWJsZS5jb20AaDdYb0pDa1p1SVdxanJIa1Q2T1RhZ3F0cDU1N1JXeGQ=</auth>
```

OUTPUT
```
#Server response
** Last event in was {xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"mechanism">>,<<"X-OAUTH2">>}],[{xmlcdata,<<"AGFkbWluQHhtcHAubWF0aW9uYWJsZS5jb20AaDdYb0pDa1p1SVdxanJIa1Q2T1RhZ3F0cDU1N1JXeGQ=">>}]}}
** When State == wait_for_feature_request
**      Data  == {state,{socket_state,fast_tls,{tlssock,#Port<0.79351>,#Port<0.79352>},<0.29719.1>},ejabberd_socket,#Ref<0.0.1048577.35198>,false,<<"15703890933737465869">>,{sasl_state,<<"jabber">>,<<"xmpp.mationable.com">>,<<>>,#Fun<ejabberd_c2s.3.32147335>,#Fun<ejabberd_c2s.4.32147335>,#Fun<ejabberd_c2s.5.32147335>,<<>>,undefined,undefined},c2s,c2s_shaper,false,true,true,true,[verify_none,compression_none,{certfile,<<"/opt/ejabberd-16.09/conf/xmpp-mationable-com.pem">>}],false,undefined,<<>>,<<"xmpp.mationable.com">>,<<>>,{{1488,425781,1736},<0.29720.1>},{pres_t,0},{pres_f,0},{pres_a,0},undefined,undefined,{userlist,none,[],false},unknown,unknown,{{124,13,11,201},10871},[],active,inactive,undefined,undefined,1000,undefined,300,300,60000,undefined,false,0,0,0,true,<<>>}
** Reason for termination =
** {function_clause,[{lists,any,[#Fun<ejabberd_oauth.5.55120481>,<<"sasl_auth">>],[{file,"lists.erl"},{line,1223}]},{cyrsasl_oauth,mech_step,2,[{file,"src/cyrsasl_oauth.erl"},{line,56}]},{cyrsasl,server_step,2,[{file,"src/cyrsasl.erl"},{line,187}]},{ejabberd_c2s,wait_for_feature_request,2,[{file,"src/ejabberd_c2s.erl"},{line,699}]},{ejabberd_c2s,decode_element,3,[{file,"src/ejabberd_c2s.erl"},{line,930}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
2017-03-02 11:36:22 =CRASH REPORT====
  crasher:
    initial call: gen:init_it/6
    pid: <0.29720.1>
    registered_name: []
    exception exit: {{function_clause,[{lists,any,[#Fun<ejabberd_oauth.5.55120481>,<<"sasl_auth">>],[{file,"lists.erl"},{line,1223}]},{cyrsasl_oauth,mech_step,2,[{file,"src/cyrsasl_oauth.erl"},{line,56}]},{cyrsasl,server_step,2,[{file,"src/cyrsasl.erl"},{line,187}]},{ejabberd_c2s,wait_for_feature_request,2,[{file,"src/ejabberd_c2s.erl"},{line,699}]},{ejabberd_c2s,decode_element,3,[{file,"src/ejabberd_c2s.erl"},{line,930}]},{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]},[{p1_fsm,terminate,8,[{file,"src/p1_fsm.erl"},{line,760}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
    ancestors: [<0.4189.0>,ejabberd_listeners,ejabberd_sup,<0.3737.0>]
    messages: []
    links: [#Port<0.79352>]
    dictionary: [{'$internal_queue_len',0}]
    trap_exit: false
    status: running
```

EXPECTED OUTPUT
```
#!xml
<success xmlns='urn:ietf:params:xml:ns:xmpp-sasl'/>
```

**Comment by zinid** [comment]
Confirmed.

**Comment by alexeyshch** [comment]
Fixed in 9a142eb8075d98b4741222bbb76f897b22795acf
